alg/teca_model_segmentation.py

```diff
@@ -0,0 +1,109 @@
+import sys
+import teca_py
+import numpy as np
+
+class teca_model_segmentation:
+    """
+    Computes summary statistics, histograms on sorted, classified,
+    TC trajectory output.
+    """
+    @staticmethod
+    def New():
+        return teca_model_segmentation()
+
+    def __init__(self):
+        self.variable_name = "IVT"
+        self.model_pt_path = None
+
+        self.impl = teca_py.teca_programmable_algorithm.New()
+        self.impl.set_number_of_input_connections(1)
+        self.impl.set_number_of_output_ports(1)
+        self.impl.set_execute_callback(self.get_tc_activity_execute(self))
+
+    def __str__(self):
+        return 'basename=%s' % (self.model_pt_path)
+
+    def set_variable_name(self, var_name):
+        """
+        set the variable name that will be inputed to the segmentation model
+        """
+        self.variable_name = var_name
+
+    def set_model_pt_path(self, model_pt_path):
+        """
+        set file path of the pretrained model that will be loaded by Pytorch
+        """
+        self.model_pt_path = model_pt_path
+
+    def set_input_connection(self, obj):
+        """
+        set the input
+        """
+        self.impl.set_input_connection(obj)
+
+    def get_output_port(self):
+        """
+        get the output
+        """
+        return self.impl.get_output_port()
+
+    def update(self):
+        """
+        execute the pipeline from this algorithm up.
+        """
+        self.impl.update()
+
+    @staticmethod
+    def get_predictions_execute(state):
+        """
+        return a teca_algorithm::execute function. a closure
+        is used to gain state.
+        """
+        def execute(port, data_in, req):
+            """
+            expects an array of an input variable to run through
+            the torch model and get the segmentation results as an 
+            output.
+            """
+            in_mesh = teca.as_teca_cartesian_mesh(data_in[0])
+
+            if in_mesh is None:
+                # TODO if this is part of a parallel pipeline then
+                # only rank 0 should report an error.
+                sys.stderr.write('ERROR: empty input, or not a mesh\n')
+                return teca_cartesian_mesh.New()
+
+            import torch
+            import torch.nn.functional as F
+
+            if self.model_pt_path is None:
+                # TODO if this is part of a parallel pipeline then
+                # only rank 0 should report an error.
+                sys.stderr.write('ERROR: pretrained model has not been specified\n')
+                return teca_cartesian_mesh.New()
+
+            model = None
```

Comment:
@burlen suggests moving lines 85-92 to the constructor or to the `set_model_pt_path()` function.

It also may be important to think about how/whether pytorch interacts with MPI: if we have 100 MPI ranks, would all try to read/load the model at the same time?  Is there a way to get pytorch to do this efficiently, such that rank 0 does the reading and broadcasts to all other ranks?

Burlen also has the idea that we might be able to store the pt file information in a header, similar to the bayesian AR detector.

Reply:
> Burlen also has the idea that we might be able to store the pt file information in a header, similar to the bayesian AR detector.

This might not be practical.  The pytorch model file that we are using is 227MB.  If we put this information into a header, that would mean that the teca executable would be at least 227MB.  Also, it isn't clear to me how easy this would be, since the .pt file is a binary, serialized version of the model parameters (see https://pytorch.org/tutorials/beginner/saving_loading_models.html).

Reply:
in that case the model file needs to be somehow packaged with teca, probably in TECA_data repo.

the pytorch documentation indicates that one can initialize the model from a "deserieralized state_dict" so we'd probably want rank 0 to load the model and broadcast this state_dict to others to avoid the I/O scalability bottleneck.

Reply:
agreed.

Reply:
the pytorch docs also say that the state_diict model representation is the recommended approach because it is smaller and more efficient than the full model. we are apparently using the full model here. I suppose that is unintentional, we probably do want the full model around for future R&D but may not need it here.

https://pytorch.org/tutorials/beginner/saving_loading_models.html#save-load-state-dict-recommended

Reply:
That's great to know.  @elbashandy - could you try re-saving the .pt file as a state-dict and see how much the file size changes?  It would be great if it were small enough that we could at least include it in the TECA repo, if not actually save it in a header somehow.

Reply:
Of course will try that!

Reply:
@burlen @taobrienlbl Reading more into it I found that we'll need to load Ankur's entire Module Python Class to do this as `state_dict` only has the model parameters but doesn't include information on the used architecture. For example:

```
# Define model
class TheModelClass(nn.Module):
    def __init__(self):
        super(TheModelClass, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

# Initialize model
model = TheModelClass()
model.load_state_dict(torch.load(PATH))
model.eval()
```

That's why state_dict is low in size. Let me know if we still should consider that option.

Reply:
that seems like that shouldn't be too big of a deal to also include a class that define's the model's structure as part of the TECA algorithm.  The other consideration would be whether we can easily code a C++ version of the same class, which would allow us to create a C++ version of the detector as we discussed.  From looking at the libtorch documentation that Burlen sent, it looks to me like that should be pretty straightforward.

What size is state_dict?

Reply:
Interesting, do you mean to create a class that constructs TheModelClass based on some parameters like the layers used and number of nodes? or do you mean to serialize the whole class in a way?

The size is still 227M

```
MODEL_NAME = "pre_trained_deeplab_resnet_IVT_artmip_probabilistic_adam_epoch_001_loss_perceptual_loss_160987.8365_time_corr_split.pt"
MODEL_DIR = os.path.join(WORK_DIR, "saved_models")
MODEL_PATH = os.path.join(MODEL_DIR, MODEL_NAME)

model = torch.load(MODEL_PATH).cuda()
model.eval()

torch.save(model.state_dict(), os.path.join(MODEL_DIR, "state_dict.pt"))
```

```
$ du -h *
227M	pre_trained_deeplab_resnet_IVT_artmip_probabilistic_adam_epoch_001_loss_perceptual_loss_160987.8365_time_corr_split.pt
227M	state_dict.pt
```

Reply:
> Interesting, do you mean to create a class that constructs TheModelClass based on some parameters like the layers used and number of nodes?

Yes, exactly, but this is apparently a moot point because of:

> The size is still 227M

It probably doesn't make sense to invest effort in coding up the class manually if it doesn't save any space; we're probably better off just simply loading the current .pt file.

Reply:
hmm, not sure what you are saying here....

> only has the model parameters but doesn't include information on the used architecture

the model is architecture independent, as it should be no?

Also, it would seem that we need to store it as a state dict, ala (`torch.save(model.state_dict(), PATH)`) even if it is the same size? the fact that it is the same size makes me wonder if what we have is already the state dict representation? At any rate...

Those issues aside, what we need is something like the following

```Python
m = MyModelClass()
if rank == 0:
  tmp = torch.load(model_path)
  comm.bcast(tmp, root=0)
else:
  tmp = comm.bcast(None, root=0)
m.load_state_dict(tmp)
```
this would prevent cost of a large number of ranks from loading the model simultaneously from killing scalability when we make very large runs

Reply:
The model parameters has to be dependent on the architecture as the parameters has to be distributed to the architecture layers correctly. For example:

```
# A linear layer that takes 120 inputs and gives out 84 outputs
# each input and output has it's own weight (model parameter) 
nn.Linear(120, 84)
```

I made sure that the 2 files are different as:
```
MODEL_NAME = "pre_trained_deeplab_resnet_IVT_artmip_probabilistic_adam_epoch_001_loss_perceptual_loss_160987.8365_time_corr_split.pt"
MODEL_DIR = os.path.join(WORK_DIR, "saved_models")
MODEL_PATH = os.path.join(MODEL_DIR, MODEL_NAME)

model = torch.load(MODEL_PATH)
model.eval()

print(model)
```
Lays out the architecture:
```
DeepLabv3_plus(
  (resnet_features): ResNet(
    (conv1): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
    (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU(inplace)
    (maxpool): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
    (layer1): Sequential(
      (0): Bottleneck(
        (conv1): Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
        (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (conv3): Conv2d(64, 256, kernel_size=(1, 1), stride=(1, 1), bias=False)
        (bn3): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (downsample): Sequential(
          (0): Conv2d(64, 256, kernel_size=(1, 1), stride=(1, 1), bias=False)
          (1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        )
      )
...
```
& 
```
torch.save(model.state_dict(), os.path.join(MODEL_DIR, "state_dict.pt"))

model = torch.load(os.path.join(MODEL_DIR, "state_dict.pt"))
print(model)
```
gives only the parameters
```
OrderedDict([('resnet_features.conv1.weight', tensor([[[[ 1.9983e-02, -6.4122e-03, -2.0461e-02,  ..., -1.5736e-02,
           -3.7028e-03,  6.0101e-03],
          [-3.7296e-03,  1.7487e-03, -1.7205e-02,  ..., -8.4745e-02,
           -7.6501e-02, -4.3329e-02],
          [ 2.6278e-03,  2.4495e-02,  4.7822e-02,  ...,  3.1007e-02,
            1.0660e-02,  2.4087e-02],
          ...,
```

Reply:
I think the confusion here might be the word 'architecture'.  In this context, I think @elbashandy is using 'architecture' to refer to the layout of the CNN, rather than the architecture of a computer system.

Reply:
@taobrienlbl Yes you are right! sorry for not clarifying this before.

Reply:
yes, I was confused on that point, totally my fault! thanks for explanation...